src/fm.py

```diff
@@ -33,5 +33,71 @@ def main():
         print(f"Search {args.genome} for {args.reads}")
 
 
+def lower(a: str, x: str, sa: memoryview, lo: int, hi: int, offset: int) -> int:
+    """Finds the lower bound of `a` at `offset` in the block defined by `lo:hi`."""
+    while lo < hi:  # Search in sa[lo:hi]
+        m = (lo + hi) // 2
+        if x[sa[m] + offset % len(x)] < a: # compare at column offset in sa
+            lo = m + 1
+        else:
+            hi = m
+    return lo
+
+def upper(a: str, x: str, sa: memoryview, lo: int, hi: int, offset: int) -> int:
+    """Finds the upper bound of `a` at `offset` in the block defined by `lo:hi`."""
+    return lower(chr(ord(a) + 1), x, sa, lo, hi, offset)
+
+def search(sa, pattern, genome):
+    lo, hi = 0, len(sa)
+    for offset, a in enumerate(pattern):
+        lo = lower(a, genome, sa, lo, hi, offset)
+        hi = upper(a, genome, sa, lo, hi, offset)
+    for sol in sa[lo:hi]:
+        yield sol+1
+    return
+
+def getSuffixes(x):
+    """
+    Gets all suffixes from a string
+    """
+    suffixes = [x[i:] for i in range(0, len(x))] 
+    # print("list of unordered suffixes: ",suffixes) 
+    return suffixes
+
+def radix_sort(lst: list[memoryview]):
+    # print("Radix sort", len(lst))
+    maximum = max(lst, key = len)
+
+    for place in range(len(maximum),0, -1):
+        lst = counting_sort(lst, place - 1)
+    
+    lst = [len(maximum) - len(suffix) for suffix in lst]
+    return lst
+            
+
+def counting_sort(lst: list[memoryview], place):
+    maximum = max(lst, key = len)
+    counts = dict.fromkeys(["$",*sorted(maximum)],[])
+    for string_index in range(len(lst)):
+        if place >= len(lst[string_index]):
+            counts["$"] = [*counts["$"], lst[string_index]]
```

Comment:
When you put stuff into a list with `[*expr, ...]` you are copying all of `expr`, so this loop can take time $O(n)$.